# Microbiome diversity of low biomass skin sites is captured by metagenomics but not 16S amplicon sequencing

Laura Markey, Evan B. Qu, Calen Mendall, Ana Finzel, Arne Materna and Tami D. Lieberman 

Established workflows for microbiome analysis work well for high microbial biomass samples, like stool, but often fail to accurately define microbial communities when applied to low microbial biomass samples. Here, we systemically compare microbiome analysis methods —16S rRNA sequencing, shallow metagenomics, and qPCR PMP™ panels—as well as extraction methods across skin swab samples and mock community dilutions. While extraction method minimally impacted results, with no significant signal for method-specific contamination or bias, we observed critical differences in inferred composition across analysis methods for low biomass samples. Metagenomic sequencing and qPCR revealed concordant, diverse microbial communities on low biomass leg skin samples, whereas 16S amplicon sequencing exhibited extreme bias toward the most abundant taxon. Both qPCR and metagenomics showed that female genital tract bacteria dominated the leg skin microbiome in about half of female subjects.  Metagenomics also enabled sub-species analysis, which demonstrated that individuals have consistent within-species diversity across high-biomass forehead and low-biomass leg skin sites. This work illustrates that shallow metagenomics provides the necessary sensitivity and taxonomic resolution to characterize species and strain-level diversity in extremely low biomass samples, opening possibilities for microbiome discovery in previously unexplored niches.

This repository contains the data and code necessary to reproduce the figures and tables shown in the publication, “Microbiome diversity of low biomass skin sites is captured by metagenomics but not 16S sequencing” with the exception of raw sequencing data which is available from NCBI BioProject ID PRJNA1277168.

## To reproduce figures from raw abundance tables:  
1.	Download the **data** and **scripts** folders  
2.	Run the script read_filter_datasets.py to generate filtered abundance tables  
3.	Run the script revised_analysis_viz.py to generate main and supplemental figures (all except Fig. S5) and tables  
4.	Run the script filter_metagenomics_methods.py to generate figures shown in Fig. S5

## To reproduce figures from raw sequencing data:  
1.	Download fastq files from NCBI BioProject PRJNA1277168  
2.	Download the **metagenomics** folder and run the Snakemake workflow to generate the bracken output used by read_filter_datasets.py  
3.	Run the script QIIME2_workflow.py to generate the raw ASV table used by read_datasets.py  
4.	Proceed from step 2 above

## Contents of the repository include:
1.	The **data** folder contains the data accessed by code throughout the repository to generate figures and tables, including:  
	-compare_filters_data: this folder includes filtered and unfiltered metagenomics relative abundance data (output of kraken) and the output of decontam as well as metadata files used to generate Fig. S5  
	-hex_color_keys: various color keys referenced by code throughout the repository to color code bacterial taxa to a hex color  
	-output_read_filter_datasets: this folder includes the output of read_filter_datasets.py  
	-PHLAME_analysis: this folder includes the output of PHLAME which was run on the unfiltered metagenomics dataset as well as associated metadata used to generate Fig. 5B and Fig. S10C-D  
	raw_data: this folder includes absolute abundance tables for qPCR analyses, relative abundance tables for metagenomic and 16S sequencing analyses, the log file describing alignment of metagenomic samples to human genome, and the classifer used for 16S classification (created by Alex Poret for the publication “Cutaneous surgical wounds have distinct microbiomes from intact skin” (10.1128/spectrum.03300-22) which is described at github.com/ajporet/cutaneous_wound_microbiome)

2.  The **metagenomics** folder contains the Snakemake workflow and associated files used to process metagenomic sequencing reads and classify them including:  
	-Snakefile: the python workflow script used for reproducible analysis  
	-scripts: a folder containing additional scripts used by this workflow  
	-envs: a folder containing files to recreate the conda environments used by this workflow  
	-samples.csv: a file used to assign reads to samples

3.	The **scripts** folder contains:  
	-Python scripts used to analyze data and create figures for this publication: filter_metagenomics_methods.py, read_filter_datasets.py and revised_analysis_viz.py  
	-A helper functions script which is called by the visualization scripts above  
	-The QIIME2 workflow used to analyze 16S sequencing data prior to the filtering and visualization scripts (QIIME2_workflow.py)

4.	The **main_figures** and **supp_figures** folders contain the raw figures generated by the code in the scripts folder. Aesthetic modifications made with Adobe Illustrator and PowerPoint to generate the figures shown in the final publication.

5.	The **main_supp_tables** folder contains CSV files for supplemental tables generated by code elsewhere in the repository




